components/brave_shields/browser/https_everywhere_service.h

```diff
@@ -43,57 +43,68 @@ struct HTTPSE_REDIRECTS_COUNT_ST {
-  explicit HTTPSEverywhereService(BraveComponent::Delegate* delegate);
+  explicit HTTPSEverywhereService(
+      scoped_refptr<base::SequencedTaskRunner> task_runner);
   HTTPSEverywhereService(const HTTPSEverywhereService&) = delete;
   HTTPSEverywhereService& operator=(const HTTPSEverywhereService&) = delete;
   ~HTTPSEverywhereService() override;
-  bool GetHTTPSURL(const GURL* url,
-                   const uint64_t& request_id,
-                   std::string* new_url);
+
+  class Engine : public base::SupportsWeakPtr<Engine> {
+   public:
+    explicit Engine(HTTPSEverywhereService* service);
+    Engine(const Engine&) = delete;
+    Engine& operator=(const Engine&) = delete;
+
+    void Init(const base::FilePath& base_dir);
+    bool GetHTTPSURL(const GURL* url,
+                     const uint64_t& request_id,
+                     std::string* new_url);
+
+   private:
+    std::string ApplyHTTPSRule(const std::string& originalUrl,
+                               const std::string& rule);
+    std::string CorrecttoRuleToRE2Engine(const std::string& to);
+    void CloseDatabase();
+
+    leveldb::DB* level_db_;
+    HTTPSEverywhereService* service_;  // not owned
+    SEQUENCE_CHECKER(sequence_checker_);
+  };
+
+  void InitDB(const base::FilePath& install_dir);
+
   bool GetHTTPSURLFromCacheOnly(const GURL* url,
                                 const uint64_t& request_id,
                                 std::string* cached_url);
 
+  base::WeakPtr<Engine> engine() { return engine_->AsWeakPtr(); }
+
  protected:
   bool Init() override;
-  void OnComponentReady(const std::string& component_id,
-      const base::FilePath& install_dir,
-      const std::string& manifest) override;
-
-  void AddHTTPSEUrlToRedirectList(const uint64_t& request_id);
-  bool ShouldHTTPSERedirect(const uint64_t& request_id);
-  std::string ApplyHTTPSRule(const std::string& originalUrl,
-      const std::string& rule);
-  std::string CorrecttoRuleToRE2Engine(const std::string& to);
 
  private:
   friend class ::HTTPSEverywhereServiceTest;
+  friend class Engine;
   static bool g_ignore_port_for_test_;
-  static std::string g_https_everywhere_component_id_;
-  static std::string g_https_everywhere_component_base64_public_key_;
   static void SetIgnorePortForTest(bool ignore);
-  static void SetComponentIdAndBase64PublicKeyForTest(
-      const std::string& component_id,
-      const std::string& component_base64_public_key);
-
-  void CloseDatabase();
 
-  void InitDB(const base::FilePath& install_dir);
+  void AddHTTPSEUrlToRedirectList(const uint64_t& request_id);
+  bool ShouldHTTPSERedirect(const uint64_t& request_id);
+  HTTPSERecentlyUsedCache<std::string>& recently_used_cache();
 
   base::Lock httpse_get_urls_redirects_count_mutex_;
   std::vector<HTTPSE_REDIRECTS_COUNT_ST> httpse_urls_redirects_count_;
   HTTPSERecentlyUsedCache<std::string> recently_used_cache_;
-  leveldb::DB* level_db_;
+  std::unique_ptr<Engine, base::OnTaskRunnerDeleter> engine_;
```

Comment:
`base::SequenceBound` is much better, also there is no need to expose WeakPtr and the task runner, just do AsynCall inside the service

Reply:
@antonok-edm we can take care of this in the follow-up PR that fixes the remaining threading issues, but please file an issue to track